In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical
#from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

ModuleNotFoundError: No module named 'tensorflow'

# LOAD DATASET

In [4]:
df=pd.read_csv('Alphabets_data.csv')
df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

# EDA

In [5]:
df.duplicated().sum()

np.int64(1332)

In [6]:
df.drop_duplicates(inplace=True)

In [8]:
df.shape

(18668, 17)

In [9]:
df.isna().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [10]:
df.describe()

xbox          ybox         width        height         onpix  \
count  18668.000000  18668.000000  18668.000000  18668.000000  18668.000000   
mean       4.150793      7.244322      5.279891      5.506910      3.660596   
std        1.879428      3.187660      1.934960      2.178292      2.160555   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        3.000000      5.000000      4.000000      4.000000      2.000000   
50%        4.000000      8.000000      5.000000      6.000000      3.000000   
75%        5.000000     10.000000      6.000000      7.000000      5.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  18668.000000  18668.000000  18668.000000  18668.000000  18668.000000   
mean       6.919381      7.529409      4.584637      5.201361      8.353171   
std        2.032226      2.319834      2.625714      2.320102      2.468905   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

             x2ybar        xy2bar         xedge        xedgey         yedge  \
count  18668.000000  18668.000000  18668.000000  18668.000000  18668.000000   
mean       6.452700      7.924577      3.131455      8.351832      3.796872   
std        2.624673      2.106957      2.323901      1.574853      2.540271   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        5.000000      7.000000      2.000000      8.000000      2.000000   
50%        6.000000      8.000000      3.000000      8.000000      4.000000   
75%        8.000000      9.000000      4.000000      9.000000      5.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

             yedgex  
count  18668.000000  
mean       7.784390  
std        1.657358  
min        0.000000  
25%        7.000000  
50%        8.000000  
75%        9.000000  
max       15.000000

# Data Exploration and Preprocessing

In [ ]:
import numpy as np

seed = 7
np.random.seed(seed)
X = df.drop(columns = ['letter'])
y = df['letter']

In [ ]:
Xtrain,Xtest,ytrain,ytest = train_test_split(X,y,test_size = 0.20, random_state=1)

In [ ]:
le = LabelEncoder()
ytrain_encoded_labels = le.fit_transform(ytrain)
ytest_encoded_labels = le.transform(ytest)

In [ ]:
from sklearn.preprocessing import StandardScaler


In [ ]:
a = StandardScaler()
a.fit(X)
X_standardized = a.transform(X)

# MODEL IMPLIMENTATIN

In [ ]:
model = Sequential()
model.add(Dense(64, input_dim=16,  activation='relu')) #1st layer
model.add(Dense(34,  activation='relu')) #2nd layer
model.add(Dense(26, activation='softmax')) #3rd layer or op layer

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
df['letter'].nunique()

26

In [ ]:
model.compile(loss='Categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
ytrain_encoded = to_categorical(ytrain_encoded_labels)
ytest_encoded = to_categorical(ytest_encoded_labels)


In [ ]:
model.compile(
    loss='categorical_crossentropy',   # for multi-class classification
    optimizer='adam',
    metrics=['accuracy']
)

In [ ]:
history = model.fit(Xtrain, ytrain_encoded, epochs=10, batch_size=32, validation_split=0.2)

Epoch 1/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 8s 12ms/step - accuracy: 0.2658 - loss: 2.4915 - val_accuracy: 0.5149 - val_loss: 1.7045
Epoch 2/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.6116 - loss: 1.3737 - val_accuracy: 0.6589 - val_loss: 1.2210
Epoch 3/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.6909 - loss: 1.0943 - val_accuracy: 0.6793 - val_loss: 1.1065
Epoch 4/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.7288 - loss: 0.9804 - val_accuracy: 0.7195 - val_loss: 0.9903
Epoch 5/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.7504 - loss: 0.9005 - val_accuracy: 0.7332 - val_loss: 0.9320
Epoch 6/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7574 - loss: 0.8466 - val_accuracy: 0.7509 - val_loss: 0.8860
Epoch 7/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7679 - loss: 0.8116 - val_accuracy: 0.7563 - val_loss: 0.8471
Epoch 8/10
374/374 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7789 - loss: 0.7666 - val_accuracy: 0

In [ ]:
print(Xtrain.shape)        # should be (rows, 16)
print(ytrain_encoded.shape)  # should be (rows, 26)


(14934, 16)
(14934, 26)


In [ ]:
# Evaluate on test set
test_loss, test_acc = model.evaluate(Xtest, ytest_encoded)
print(f"Test Accuracy: {test_acc:.4f}")


117/117 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7852 - loss: 0.7219
Test Accuracy: 0.7852


In [ ]:
from sklearn.metrics import classification_report
import numpy as np

# Predict classes
y_pred_probs = model.predict(Xtest)             # probabilities
y_pred_classes = np.argmax(y_pred_probs, axis=1) # predicted labels
y_true = np.argmax(ytest_encoded, axis=1)        # true labels

# Classification report
print(classification_report(y_true, y_pred_classes))


117/117 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
              precision    recall  f1-score   support

           0       0.78      0.93      0.85       153
           1       0.63      0.81      0.71       138
           2       0.73      0.71      0.72       133
           3       0.69      0.90      0.78       155
           4       0.57      0.85      0.68       131
           5       0.84      0.73      0.78       153
           6       0.70      0.47      0.56       131
           7       0.71      0.54      0.61       131
           8       0.82      0.70      0.76       106
           9       0.88      0.89      0.88       159
          10       0.53      0.73      0.62       129
          11       0.74      0.89      0.81       132
          12       0.92      0.86      0.89       151
          13       0.90      0.83      0.86       156
          14       0.86      0.67      0.75       154
          15       0.82      0.86      0.84       175
          16       0.86      0.75      0

In [ ]:
# HYPERPARAMETER TUNING

In [ ]:
#!pip install scikeras

In [ ]:
#!pip uninstall scikeras tensorflow keras scikit-learn -y


In [ ]:
'''!pip uninstall -y scikeras tensorflow keras scikit-learn
!pip install tensorflow==2.17.1 scikeras==0.13.0 scikit-learn==1.5.2'''


In [ ]:
######################

In [23]:
# --- Step 1: Imports ---
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import RandomizedSearchCV
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import accuracy_score, classification_report
import numpy as np

# --- Step 2: Build model function ---
def create_model(learning_rate=0.001, activation='relu', neurons=32):
    model = Sequential()
    model.add(Dense(neurons, input_dim=Xtrain.shape[1], activation=activation))
    model.add(Dense(neurons, activation=activation))
    model.add(Dense(ytrain_encoded.shape[1], activation='softmax'))  # for multi-class classification
    optimizer = Adam(learning_rate=learning_rate)
    model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

# --- Step 3: Wrap model ---
model = KerasClassifier(model=create_model, verbose=0)

# --- Step 4: Define parameter grid ---
param_grid = {
    "batch_size": [16, 32],
    "epochs": [10, 20],
    "model__learning_rate": [0.001, 0.01],
    "model__activation": ["relu", "tanh"],
    "model__neurons": [16, 32]
}

# --- Step 5: Randomized Search (faster than GridSearchCV) ---
random_search = RandomizedSearchCV(
    estimator=model,
    param_distributions=param_grid,
    n_iter=5,   # test only 5 random combinations
    cv=3,
    n_jobs=-1,
    verbose=1
)

# --- Step 6: Fit the model ---
random_result = random_search.fit(Xtrain, ytrain_encoded)

# --- Step 7: Display best results ---
print("✅ Best Parameters Found:", random_result.best_params_)
print("✅ Best Cross-Validation Accuracy:", random_result.best_score_)


Fitting 3 folds for each of 5 candidates, totalling 15 fits


/usr/local/lib/python3.12/dist-packages/numpy/ma/core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,
/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


✅ Best Parameters Found: {'model__neurons': 32, 'model__learning_rate': 0.001, 'model__activation': 'tanh', 'epochs': 20, 'batch_size': 16}
✅ Best Cross-Validation Accuracy: 0.8396946564885496


# Evaluate Tuned Model

In [24]:
# Evaluate tuned model
best_model = random_result.best_estimator_
y_pred_tuned = best_model.predict(Xtest)

# Convert predictions back to class labels if one-hot encoded
y_pred_tuned = np.argmax(y_pred_tuned, axis=1)
y_true = np.argmax(ytest_encoded, axis=1)

print("✅ Accuracy (Tuned Model):", accuracy_score(y_true, y_pred_tuned))
print("\n✅ Classification Report:\n", classification_report(y_true, y_pred_tuned))


✅ Accuracy (Tuned Model): 0.8668987680771291

✅ Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.90      0.93       153
           1       0.89      0.80      0.84       138
           2       0.90      0.70      0.79       133
           3       0.85      0.95      0.90       155
           4       0.88      0.75      0.81       131
           5       0.89      0.77      0.83       153
           6       0.71      0.84      0.77       131
           7       0.78      0.71      0.74       131
           8       0.90      0.86      0.88       106
           9       0.90      0.92      0.91       159
          10       0.82      0.82      0.82       129
          11       0.89      0.89      0.89       132
          12       0.95      0.93      0.94       151
          13       0.90      0.90      0.90       156
          14       0.78      0.86      0.82       154
          15       0.87      0.89      0.88       175
         

### Performance Discussion



| Model Type | Accuracy | Comments |
|-------------|-----------|-----------|
| Default Model | 0.79 | Basic architecture without optimization. |
| Tuned Model | 0.87 | Improved accuracy after adjusting learning rate, neurons, and activation function. |

**Observation:**
After hyperparameter tuning, the model achieved better accuracy and generalization.
This shows that tuning parameters like learning rate and number of neurons significantly affects training stability and performance.
